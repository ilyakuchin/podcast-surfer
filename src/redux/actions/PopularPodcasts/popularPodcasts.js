import axios from 'axios';
import { popularPodcastsUrl } from '../../../helpers/urls';

export const FETCH_POPULAR_PODCASTS_REQUEST = 'FETCH_POPULAR_PODCASTS_REQUEST';
export const FETCH_POPULAR_PODCASTS_SUCCESS = 'FETCH_POPULAR_PODCASTS_SUCCESS';
export const FETCH_POPULAR_PODCASTS_FAILURE = 'FETCH_POPULAR_PODCASTS_FAILURE';

export function fetchPopularPodcastsRequest() {
  return { type: FETCH_POPULAR_PODCASTS_REQUEST, isFetching: true };
}

export function fetchPopularPodcastsSuccess(popularPodcasts) {
  return {
    type: FETCH_POPULAR_PODCASTS_SUCCESS,
    isFetching: false,
    popularPodcasts
  };
}

export function fetchPopularPodcastsFailure(error) {
  return {
    type: FETCH_POPULAR_PODCASTS_FAILURE,
    isFetching: false,
    error
  };
}

export function fetchPopularPodcasts() {
  return dispatch => {
    dispatch(fetchPopularPodcastsRequest());
    return axios
      .get(popularPodcastsUrl)
      .then(res => {
        dispatch(fetchPopularPodcastsSuccess(res.data));
      })
      .catch(err => {
        dispatch(fetchPopularPodcastsFailure(err));
      });
  };
}
