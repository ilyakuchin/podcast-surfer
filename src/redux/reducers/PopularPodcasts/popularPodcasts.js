import {
  FETCH_POPULAR_PODCASTS_REQUEST,
  FETCH_POPULAR_PODCASTS_SUCCESS,
  FETCH_POPULAR_PODCASTS_FAILURE
} from '../../actions/PopularPodcasts/popularPodcasts';

export default function podcasts(
  state = { isFetching: false, podcasts: [], error: '' },
  action
) {
  switch (action.type) {
    case FETCH_POPULAR_PODCASTS_REQUEST:
      return { ...state, isFetching: action.isFetching };
    case FETCH_POPULAR_PODCASTS_SUCCESS:
      return {
        ...state,
        isFetching: action.isFetching,
        podcasts: action.popularPodcasts
      };
    case FETCH_POPULAR_PODCASTS_FAILURE:
      return {
        ...state,
        isFetching: action.isFetching,
        error: action.error
      };
    default:
      return state;
  }
}
