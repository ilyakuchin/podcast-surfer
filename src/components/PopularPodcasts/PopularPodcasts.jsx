import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import ConnectedPodcastList from '../PodcastList/PodcastList';
import { fetchPopularPodcasts } from '../../redux/actions/PopularPodcasts/popularPodcasts';

function PopularPodcasts({
  fetchPopularPodcastsConnect,
  popularPodcasts,
  isFetching
}) {
  useEffect(() => {
    fetchPopularPodcastsConnect();
  }, [fetchPopularPodcastsConnect]);

  return (
    <ConnectedPodcastList podcasts={popularPodcasts} isFetching={isFetching} />
  );
}

PopularPodcasts.propTypes = {
  fetchPopularPodcastsConnect: PropTypes.func.isRequired,
  isFetching: PropTypes.bool.isRequired,
  popularPodcasts: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      image: PropTypes.string,
      rss: PropTypes.string,
      name: PropTypes.string
    })
  ).isRequired
};

const mapStateToProps = state => {
  return {
    popularPodcasts: state.popularPodcasts.podcasts,
    isFetching: state.popularPodcasts.isFetching
  };
};

const dispatchToProps = dispatch => {
  return {
    fetchPopularPodcastsConnect: () => dispatch(fetchPopularPodcasts())
  };
};

export default connect(mapStateToProps, dispatchToProps)(PopularPodcasts);
