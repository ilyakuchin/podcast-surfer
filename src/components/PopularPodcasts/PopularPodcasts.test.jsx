import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import { configure, shallow } from 'enzyme';
import configureStore from 'redux-mock-store';
import toJson from 'enzyme-to-json';
import PopularPodcasts from './PopularPodcasts';
import mockState from '../../helpers/index';

const mockStore = configureStore([]);
configure({ adapter: new Adapter() });

let store;
let component;

describe('Test PopularPodcasts component', () => {
  beforeEach(() => {
    store = mockStore({
      ...mockState,
      fetchPopularPodcasts: () => 'fetchPopularPodcasts'
    });

    component = shallow(<PopularPodcasts store={store} />);
  });

  test('should render correctly', () => {
    expect(toJson(component)).toMatchSnapshot();
  });
});
