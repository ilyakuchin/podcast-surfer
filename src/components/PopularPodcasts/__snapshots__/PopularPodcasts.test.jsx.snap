// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test PopularPodcasts component should render correctly 1`] = `
<ContextProvider
  value={null}
>
  <PopularPodcasts
    fetchPopularPodcastsConnect={[Function]}
    isFetching={false}
    popularPodcasts={
      Array [
        Object {
          "id": "1",
          "image": "toplink",
          "name": "topname",
          "rss": "toprss1",
        },
        Object {
          "id": "2",
          "image": "toplink2",
          "name": "topname2",
          "rss": "toprss2",
        },
        Object {
          "id": "3",
          "image": "toplink3",
          "name": "topname3",
          "rss": "toprss3",
        },
      ]
    }
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  />
</ContextProvider>
`;
